### ISSUE TITLE
False negatives in (some) compressed PKZIP files

### ISSUE THREAD

**Issue by hops** [hashcat/hashcat#3434]
hashcat fails to crack certain compressed PKZIP files. I think the condition is if the uncompressed size is larger than 32768 bytes. More testing is certainly needed in that area!
This was [reported](https://discord.com/channels/566331281558732800/874753308953280613/1008755855681929268) by 64_nickel on Discord.

To reproduce I've attached the mentioned [list16-FL_kdIZUGpI.zip](https://github.com/hashcat/hashcat/files/9451947/list16-FL_kdIZUGpI.zip) file (and the with zip2john extracted hash [list16-FL_kdIZUGpI.txt](https://github.com/hashcat/hashcat/files/9451946/list16-FL_kdIZUGpI.txt)) from the CMIYC 2022 street hashes. The password is `Hackers`
```
$ echo Hackers | ./hashcat --username -m 17200 list16-FL_kdIZUGpI.txt
hashcat (v6.2.5-667-g7ca2627bc+) starting
<--snip-->
Session..........: hashcat
Status...........: Exhausted
Hash.Mode........: 17200 (PKZIP (Compressed))
Hash.Target......: $pkzip$1*1*2*0*6c59*b7ae*77d34619*0*50*8*6c59*48d4*...pkzip$
<--snip-->
```
This bug was introduced with #2970. While I think we do want to have that additional check, we can't rely on `infstream.total_out` because it gets [reset](https://github.com/hashcat/hashcat/blob/master/OpenCL/inc_zip_inflate.cl#L1210) in `hc_inflate()` if `pStream->avail_out` falls below `TINFL_LZ_DICT_SIZE`.

I've mainly opened this issue because I don't know what the best solution is. We could add a separate counter to track the uncompressed bytes but maybe there's a better/easier solution?

**Comment by Chick3nman** [comment]
Could this be related to https://github.com/hashcat/hashcat/issues/3277 ?

**Comment by x151973** [comment]
The zip file above I tested can be solved with john, but hashcat failed(Device #1: Skipping (hash-mode 17200)......)
From my other tests, hashcat always failed on compressed zip file : Device #1: Skipping (hash-mode 17200......) ,except uncompressed ones(trim hashtxt begin with `$pkzip$`)
My PC is amd cpu with intel opencl driver installed, no gpu driver, hashcat -I has no error

**Comment by s3inlc** [comment]
> The zip file above I tested can be solved with john, but hashcat failed(Device https://github.com/hashcat/hashcat/issues/1: Skipping (hash-mode 17200)......)
From my other tests, hashcat always failed on compressed zip file : Device https://github.com/hashcat/hashcat/issues/1: Skipping (hash-mode 17200......) ,except uncompressed ones(trim hashtxt begin with $pkzip$)
My PC is amd cpu with intel opencl driver installed, no gpu driver, hashcat -I has no error

You must have another issue with your setup / driver etc.. The hashes still should be normally loaded (included the one posted above). The only reason where it rejects hashes is when they are too long (>320KB).

**Comment by s3inlc** [comment]
> Could this be related to #3277 ?

Most likely not, there are different reasons for these cases.

**Comment by x151973** [comment]
> > The zip file above I tested can be solved with john, but hashcat failed(Device #1: Skipping (hash-mode 17200)......)
> > From my other tests, hashcat always failed on compressed zip file : Device #1: Skipping (hash-mode 17200......) ,except uncompressed ones(trim hashtxt begin with pkzip)
> > My PC is amd cpu with intel opencl driver installed, no gpu driver, hashcat -I has no error
> 
> You must have another issue with your setup / driver etc.. The hashes still should be normally loaded (included the one posted above). The only reason where it rejects hashes is when they are too long (>320KB).

Can you show me full cli command that you successfully cracked using hash file above list16-FL_kdIZUGpI.txt?
And what platform are you using?

UPDATE:
Adding --force ignore the skipping error, but opencl error raised
Downgrade intel opencl driver, finally, the 17xxx series seems fully worked on my ryzen, cpu only machine. (hashcat -b -m 17xxx)

But, yes, some kind of zips still cannot cracked, including title one, even give password explicitly

And, I found zip2john gave different result between windows and linux(kali): pkzip, or pkzip2 and follwing short digits....

UPDATE2:
Tried a practical zip package with old time created, multi files, store, size20MB
On windows zip2john gave pkzip2.... hashcat 17225 worked, 17230 result  SAME length WRONG key
On kali zip2john gave pkzip.... hashcat both 17225 and 17230 worked

**Comment by meltingscales** [comment]
I think I might be dealing with the same issue.

    hashcat -m 1500 -a 0 secret.hash /usr/share/wordlists/rockyou.txt

secret.hash:
```
$pkzip$1*1*2*0*14*8*77537827*0*24*0*14*7753*8061b9caf8436874ad47a9481863b54443379d4c*$/pkzip$
```

**Comment by meltingscales** [comment]
Trying to do HTB "Poison" box the hard way, I wanted to test my hashcat setup but it seems I cannot use this hash after using `https://github.com/hashstation/zip2hashcat/tree/main` to extract the hash.

**Comment by huhu415** [comment]
I also have a file named `hash.txt` starting with `$pkzip$` (https://github.com/user-attachments/files/25458913/hash.txt), which cannot be cracked by hashcat. However, I know the password, which is 106836.

This `hash.txt` file is from zip2john.

``` bash

❯ ./run/zip2john /Users/hello/Downloads/bank.zip > /dev/null

ver 2.0 bank.zip/xxxxxxxxx.pdf PKZIP Encr: TS_chk, cmplen=374904, decmplen=1088216, crc=ECA9D54C ts=7D7C cs=7d7c type=8

```

How can I solve this problem?

**Comment by meltingscales** [comment]
Maybe the `.zip` extension file specification allows for different password hash formats? Do we need to dive into the `.zip` spec to find out the core of this issue?

**Comment by meltingscales** [comment]
OKAY, so, I asked Claude Code, so sorry for slop-posting, but- this might be fixed and I just need to update hashcat! Maybe Kali devs need to merge it into main??

# Claude Response

  hashcat's PKZIP compressed mode (-m 17200) has a bug where it fails to crack hashes when the uncompressed data > 32KB (32768 bytes). The bug is in how infstream.total_out gets reset      
  during decompression in hc_inflate().                                                                                                                                                      

  Key Things to Investigate

  1. Check your hashcat version
  ./hashcat --version
  The fix was merged in PR #3547 on Dec 22, 2022. If you're on a version before that, update.

  2. Verify the correct hash mode
  Looking at the thread, meltingscales made an error — they used -m 1500 (DES/crypt) instead of -m 17200 (PKZIP Compressed). Check:
  - -m 17200 — PKZIP (Compressed)
  - -m 17210 — PKZIP (Uncompressed)
  - -m 17220 — PKZIP (Mixed Multi-File)

  3. Check hash source — zip2john vs zip2hashcat
  zip2john output is not directly compatible with hashcat. You need to use a converter:
  # Use zip2hashcat or hashcat's own extraction utility
  # NOT zip2john directly

  4. Check the uncompressed size in the hash
  For huhu415's case:
  cmplen=374904, decmplen=1088216
  decmplen=1088216 (~1MB) is well over 32KB — this is exactly the bug. Confirm the fix is present in their hashcat build.

  5. Look at the relevant source code

  The bug lives in the inflate/decompression logic. In the hashcat source:
  - src/modules/module_17200.c (or similar)
  - Look for hc_inflate() and infstream.total_out
  - PR #3547 added a separate counter to track total uncompressed bytes across multiple inflate() calls

  Quick Test

  # After updating hashcat, test with known password to confirm fix:
  echo "106836" | ./hashcat -m 17200 hash.txt --stdin
  # Should show "Cracked" not "Exhausted"

  Summary

  The root cause: zlib's total_out field only tracks bytes output in the current stream window, not the total across all calls. When decompressed data exceeds TINFL_LZ_DICT_SIZE (32768),
  the stream resets and total_out restarts from 0, causing the size check introduced in PR #2970 to incorrectly reject valid candidates. PR #3547 fixed this by maintaining a separate
  running total.

**Comment by huhu415** [comment]
> echo "106836" | ./hashcat -m 17200 hash.txt --stdin

I actually used the latest version of hashcat - v7.1.2

Error:
```Hashfile 'hash.txt' on line 1 ($pkzip...b65628bf4c9ac950e3f314b*$/pkzip$): Token length exception

* Token length exception: 1/1 hashes

This error happens if the wrong hash type is specified, if the hashes are

malformed, or if input is otherwise not as expected (for example, if the

--username or --dynamic-x option is used but no username or dynamic-tag is present)

No hashes loaded.

```

I think it's a problem that `zip2john output is not directly compatible with hashcat`.

Because hashcat supports `$pkzip2$`. My input is `$pkzip$`.
